pkg/extensions/tbtc/tbtc.go

```diff
@@ -97,6 +134,270 @@ func (t *tbtc) monitorRetrievePubKey(
 	return nil
 }
 
+func (t *tbtc) monitorProvideRedemptionSignature(
+	ctx context.Context,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) error {
+	monitoringStartFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		// Start right after a redemption has been requested or the redemption
+		// fee has been increased.
+		return t.chain.OnDepositRedemptionRequested(handler)
+	}
+
+	monitoringStopFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		// Stop in case the redemption signature has been provided by someone else.
+		signatureSubscription, err := t.chain.OnDepositGotRedemptionSignature(handler)
+		if err != nil {
+			return nil, err
+		}
+
+		// Stop in case the redemption proof has been provided by someone else.
+		redeemedSubscription, err := t.chain.OnDepositRedeemed(handler)
+		if err != nil {
+			return nil, err
+		}
+
+		return subscription.NewEventSubscription(
+			func() {
+				signatureSubscription.Unsubscribe()
+				redeemedSubscription.Unsubscribe()
+			},
+		), nil
+	}
+
+	actFn := func(depositAddress string) error {
+		keepAddress, err := t.chain.KeepAddress(depositAddress)
+		if err != nil {
+			return err
+		}
+
+		redemptionRequestedEvents, err := t.chain.PastDepositRedemptionRequestedEvents(
+			depositAddress,
+			0, // TODO: Should be something better that 0
```

Comment:
This could make all clients deployed against Infura unhappy. As far as I know, they allow to reach no more than 1000 blocks in the past and assuming each block takes 15sec on average to mine, this is just 250 minutes back.

If there is no workaround, without seriously sacrificing code quality, we are fine. Just raising a concern.

Reply:
Unfortunately, the approach presented here is the safest and simplest one. We can cache incoming events as an alternative but given event duplications, chain reorgs and stuff, I want to avoid caching as it strongly complicates the code.

Reply:
I've added the link to this discussion to the [TODOs section](https://github.com/keep-network/keep-ecdsa/pull/585#discussion_r513447505) where I listed all the things related to events deduplication, reorgs etc.

Reply:
Leaving this discussion open until all follow-up PRs addressing those issues are reviewed and merged.